blog post for KEP-3619: Fine-grained SupplementalGroups control graduates to beta (#49944)
kubernetes/website: content/en/blog/_posts/XXXX-XX-XX-fine-grained-supplementalgroups-control-beta/index.md

Reviewer: Yes, indeed we will adjust this on our side, as the date is yet to be confirmed and the YYYYMMDD in the file name would also need to be updated 👍

Author: Thanks! I leave this as is, then 🙇.

---
layout: blog
title: 'Kubernetes 1.33: Fine-grained SupplementalGroups control graduates to beta'
date: XXXX-XX-XX
slug: fine-grained-supplementalgroups-control-beta
author: >
  Shingo Omura (LY Corporation)

---

Kubernetes 1.31 introduced new field `SupplementalGroupsPolicy` in Pod's Security Context to improve supplemental groups handling in containers in Pods.

In Kubernetes 1.33, the feature graduates to beta and the corresponding feature gate (`SupplementalGroupsPolicy`) gets enabled by default.

> Please be aware that this beta release contains some behavioral breaking change. See [The Behavioral Changes Introduced In Beta](#the-behavioral-changes-introduced-in-beta) and [Upgrade Considerations](#upgrade-consideration) sections for details.

Reviewer 2: do you have the same sections in documentation? It is valuable to have them in docs, not only in a blog post

Author: Basically yes. I added the behaviral changes as a note:

https://github.com/kubernetes/website/blob/7a9e33eb515a1afc2005fdef2127a701ef3548bb/content/en/docs/tasks/configure-pod-container/security-context.md?plain=1#L339-L356

But, I did not include "upgrade consideration" part in the doc. Would you think we should include it in the doc, too??

## Motivation: Implicit group memberships defined in `/etc/group` in the container image

> Although this behavior may not be popular with many Kubernetes cluster users/admins, kubernetes, by default, _merges_ group information from the Pod with information defined in `/etc/group` in the container image.

Reviewer 3: I don't know if `popular` is the correct wording here. maybe `Although the majority of Kubernetes cluster admins/users may not be aware, kubernetes, by default...`

Author: fixed in [caff013](https://github.com/kubernetes/website/pull/49944/commits/caff01322db36f60553e11d26aff2e78916f8ab9)

Let's see an example, below Pod specifies `runAsUser=1000`, `runAsGroup=3000` and `supplementalGroups=4000` in the Pod's security context.

> {{% code_sample file="implicit-groups.yaml" %}}

Author: fixed in [1a307da](https://github.com/kubernetes/website/pull/49944/commits/1a307da81b31428d8da1d6071bcccf77ad84c175)

What is the result of `id` command in the `ctr` container?

```console
# Create the Pod:

> $ kubectl apply -f https://k8s.io/blog/XXXX/XX/XX/Fine-grained-SupplementalGroups-control/implicit-groups.yaml

Reviewer 4: We don't recommend encouraging people to run `kubectl apply` to run manifests from blog articles; in fact, I advise against it.

### What's wrong with it?

The _implicitly_ merged group information from `/etc/group` in the container image may cause some concerns particularly in accessing volumes (see [kubernetes/kubernetes#112879](https://issue.k8s.io/112879) for details) because file permission is controlled by uid/gid in Linux. Even worse, the implicit gids from `/etc/group` can not be detected/validated by any policy engines because there is no clue for the implicit group information in the manifest. This can also be a concern for Kubernetes security.

## Fine-grained SupplementalGroups control in a Pod: `SupplementaryGroupsPolicy`

To tackle the above problem, Pod's `.spec.securityContext` now have `supplementalGroupsPolicy` field.

This field provides a way to control how to calculate supplementary groups for the container processes in a Pod. The available policy is below:

* _Merge_: The group membership defined in `/etc/group` for the container's primary user will be merged. If not specified, this policy will be applied (i.e. as-is behavior for backward compatibility).

> * _Strict_: it only attaches specified group IDs in `fsGroup`, `supplementalGroups`, or `runAsGroup` fields as the supplementary groups of the container processes. This means no group membership defined in `/etc/group` for the container's primary user will be merged.

Reviewer 2: should we say that this is a recommended option and we hope all Pod will use it in a few years?

Author: Yes in terms of security, personally. Probably, bunch of cluster admins does not aware the "implicit" gids from the container images.

Although it is not strong recommendation, docs for this feature includes a pointer to the risk of _implicitly merged_ gids:

https://github.com/kubernetes/website/blob/7a9e33eb515a1afc2005fdef2127a701ef3548bb/content/en/docs/tasks/configure-pod-container/security-context.md?plain=1#L227-L231

Reviewer: Making sure the sentence follows the above "Merge" policy, and use "ignored" to
make it clear that this will remove some default user setup

```suggestion
* _Strict_: Only the group IDs specified in `fsGroup`, `supplementalGroups`, or `runAsGroup` are attached as supplementary groups to the container processes. Group memberships defined in `/etc/group` for the container's primary user are ignored.
```

Let's see how `Strict` policy works.

{{% code_sample file="strict-supplementalgroups-policy.yaml" %}}

```console
# Create the Pod:

> $ kubectl apply -f https://k8s.io/blog/XXXX/XX/XX/Fine-grained-SupplementalGroups-control/strict-supplementalgroups-policy.yaml

Reviewer 2: will this link actually work? curious

Reviewer 2: Also it may be beneficial to have the source of it listed in a blog post. I wonder if a blog engine here supports the expandable code listings that are collapsed by default. This will be useful here

Author: > will this link actually work? curious

Yes. https://k8s.io/blog/2024/08/22/Fine-grained-SupplementalGroups-control/implicit-groups.yaml (note: this is a link for alpha feature blog) does work.

> Also it may be beneficial to have the source of it listed in a blog post.

It already does. [L.75](https://github.com/kubernetes/website/pull/49944/files#diff-32d533646b2ddbbbeef5d2b061deb05cf012cc6c811326ea9e2f45827433b0afR75) shows the content of `strict-supplementalgroups-policy.yaml`. [Here](https://deploy-preview-49944--kubernetes-io-main-staging.netlify.app/blog/1/01/01/fine-grained-supplementalgroups-control-beta/#:~:text=supplementalgroups%2Dpolicy.yaml-,apiVersion%3A%20v1,allowPrivilegeEscalation%3A%20false,-%23%20Create%20the) is a preview.

> I wonder if a blog engine here supports the expandable code listings that are collapsed by default. This will be useful here

Yeah, but, I'm not sure about this. Could you refer some person to whom I can ask this??

Reviewer 4: We don't recommend encouraging people to run `kubectl apply` to run manifests from blog articles; in fact, I advise against it.

{{</note>}}

## Attached process identity in Pod status

This feature also exposes the process identity attached to the first container process of the container
via `.status.containerStatuses[].user.linux` field. It would be helpful to see if implicit group IDs are attached.

```yaml
...
status:
  containerStatuses:
  - name: ctr
    user:
      linux:
        gid: 3000
        supplementalGroups:
        - 3000
        - 4000
        uid: 1000
...
```

{{<note>}}
Please note that the values in `status.containerStatuses[].user.linux` field is _the firstly attached_
process identity to the first container process in the container. If the container has sufficient privilege
to call system calls related to process identity (e.g. [`setuid(2)`](https://man7.org/linux/man-pages/man2/setuid.2.html), [`setgid(2)`](https://man7.org/linux/man-pages/man2/setgid.2.html) or [`setgroups(2)`](https://man7.org/linux/man-pages/man2/setgroups.2.html), etc.), the container process can change its identity. Thus, the _actual_ process identity will be dynamic.
{{</note>}}

## `Strict` Policy requires newer CRI versions

> Actually, CRI runtime(e.g. containerd, CRI-O) plays a core role for calculating supplementary group ids to be attached to the containers. Thus, `SupplementalGroupsPolicy=Strict` requires the CRI runtime that support this feature (`SupplementalGroupsPolicy: Merge` can work with the CRI runtime which does not support this feature because this policy is fully backward compatible policy).

Reviewer 3: nit: space between runtime and (

Author: fixed in [1a307da](https://github.com/kubernetes/website/pull/49944/commits/1a307da81b31428d8da1d6071bcccf77ad84c175)

> Actually, CRI runtime(e.g. containerd, CRI-O) plays a core role for calculating supplementary group ids to be attached to the containers. Thus, `SupplementalGroupsPolicy=Strict` requires the CRI runtime that support this feature (`SupplementalGroupsPolicy: Merge` can work with the CRI runtime which does not support this feature because this policy is fully backward compatible policy).

Author: fixed in [1a307da](https://github.com/kubernetes/website/pull/49944/commits/1a307da81b31428d8da1d6071bcccf77ad84c175)

> Actually, CRI runtime (e.g. containerd, CRI-O) plays a core role for calculating supplementary group ids to be attached to the containers. Thus, `SupplementalGroupsPolicy=Strict` requires a CRI runtime that support this feature (`SupplementalGroupsPolicy: Merge` can work with the CRI runtime which does not support this feature because this policy is fully backward compatible policy).

Reviewer 4: How about:
```suggestion
Actually, the container runtime (e.g. containerd, CRI-O) plays a core role for calculating supplementary group ids to be attached to the containers.
Thus, `supplementalGroupsPolicy: Strict` requires a CRI-compatbile container runtime
that support thiss feature. By contrast, `supplementalGroupsPolicy: Merge` can work with a container that which does not support supplementary group control,  because this policy represents the legacy behavior.
```

Author: > CRI-compatbile container runtime

kubelet always talks to the container runtime via CRI, so the original sentence calls it CRI runtime. The version of the container runtime is important here, not the compatibility. So, How about this??

```suggestion
Actually, the container runtime (e.g. containerd, CRI-O) plays a core role for calculating supplementary group ids to be attached to the containers.
Thus, `supplementalGroupsPolicy: Strict` requires the container runtime that supports this feature. 
By contrast, `supplementalGroupsPolicy: Merge` can work with a container runtime which does not support supplementary group control, because this policy represents the legacy behavior.
```

Below list is the popular CRI runtimes which supports this feature:

- containerd: v2.0 or later
- CRI-O: v1.31 or later

And, you can see if the feature is supported in the Node's `.status.features.supplementalGroupsPolicy` field.

```yaml
apiVersion: v1
kind: Node
...
status:
  features:
    supplementalGroupsPolicy: true
```

## The Behavioral Changes Introduced In Beta

> At alpha release, when a pod with `SupplementalGroupsPolicy=Strict` are scheduled to a node that does NOT support this feature(i.e. `.status.features.supplementalGroupsPolicy=false`), the pod's supplemental groups policy gets fallback-ed to the `Merge` policy _silently_.

Reviewer 3: s/gets fallback-ed/falls back`

Author: fixed in [1a307da](https://github.com/kubernetes/website/pull/49944/commits/1a307da81b31428d8da1d6071bcccf77ad84c175)

* _Strict_: Only the group IDs specified in `fsGroup`, `supplementalGroups`, or `runAsGroup` are attached as supplementary groups to the container processes. Group memberships defined in `/etc/group` for the container's primary user are ignored.

Let's see how `Strict` policy works. Below Pod manifest specifies `supplementalGroupsPolicy: Strict`:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: strict-supplementalgroups-policy
spec:
  securityContext:
    runAsUser: 1000
    runAsGroup: 3000
    supplementalGroups: [4000]
    supplementalGroupsPolicy: Strict
  containers:
  - name: ctr
    image: registry.k8s.io/e2e-test-images/agnhost:2.45
    command: [ "sh", "-c", "sleep 1h" ]
    securityContext:
      allowPrivilegeEscalation: false
```

The result of `id` command in the `ctr` container should be similar to this:

```none
uid=1000 gid=3000 groups=3000,4000
```

You can see `Strict` policy can exclude group `50000` from `groups`! 

Thus, ensuring `supplementalGroupsPolicy: Strict` (enforced by some policy mechanism) helps prevent the implicit supplementary groups in a Pod.

{{<note>}}

> Keep in mind: a container with sufficient privileges can still change its process identity. The `supplementalGroupsPolicy` only affect the initial process identity. See the following section for details.

Reviewer 5: ```suggestion
A container with sufficient privileges can change its process identity. The `supplementalGroupsPolicy` only affect the initial process identity. See the following section for details.
```

It improve readability while making it direct and simple

---
layout: blog
title: 'Kubernetes 1.33: Fine-grained SupplementalGroups control graduates to beta'

> date: XXXX-XX-XX

Reviewer: While this blog is about the v1.33 release, we will limit to one blog published per day. This means it is likely that this blog will be planned on a later date, and the schedule is yet to be fully decided. (It is highly dependent on other blog reviews completed in time)

If you can simply add `draft: true` in the front matter, it would allow us to merge the blog without actually publishing them to the website. Once the blog content is finalised and merged, the Release Comms team can take over the rest of the publication setup.

```suggestion
date: 2025-04-23
draft: true
```

You can find more in the blog guideline [here](https://kubernetes.io/docs/contribute/blog/article-submission/#article-scheduling)! 👍

Author: I see! Thank you so much! I applied your suggestion.

---
layout: blog

> title: 'Kubernetes 1.33: Fine-grained SupplementalGroups control graduates to beta'

Reviewer: We should update this based on https://kubernetes.io/docs/contribute/style/style-guide/#headings (and also version number should be prefixed with "v"):

```suggestion
title: 'Kubernetes v1.33: Fine-grained SupplementalGroups Control Graduates to Beta'
```

date: 2025-04-23
draft: true
slug: fine-grained-supplementalgroups-control-beta
author: >
  Shingo Omura (LY Corporation)

---

> Pod supplemental groups policy control for Pods is now beta. Read on to learn how it will help your cluster security, and how to start using it.

Reviewer: This is a bit generic, and the latter sentence can be completely dropped. Perhaps we could highlight how this is about the "fine-grained" control provided for better security, and combine it with the sentence below?

Author: Thanks for this suggestion. It makes sense. How about [this](tbd)??

Author: fixed in [f403e82](https://github.com/kubernetes/website/pull/49944/commits/f403e82b11d9e170122a4222cd4cb2be21bf3c8e)

> The new field was introduced as an opt-in alpha feature for Kubernetes 1.31 and has graduated to beta in 1.33; the corresponding feature gate (`SupplementalGroupsPolicy`) is now enabled by default.

Reviewer: Introduce the field detail early to let the readers know what this blog is about

```suggestion
The new field, `supplementalGroupsPolicy`, was introduced as an opt-in alpha feature for Kubernetes 1.31 and has graduated to beta in 1.33; the corresponding feature gate (`SupplementalGroupsPolicy`) is now enabled by default.
```

> Pod supplemental groups policy control for Pods is now beta. Read on to learn how it will help your cluster security, and how to start using it.
>  
> The new field, `supplementalGroupsPolicy`, was introduced as an opt-in alpha feature for Kubernetes 1.31 and has graduated to beta in 1.33; the corresponding feature gate (`SupplementalGroupsPolicy`) is now enabled by default.

Author: ```suggestion
The new field, `supplementalGroupsPolicy`, was introduced as an opt-in alpha feature for Kubernetes 1.31 and has graduated to beta in 1.33; the corresponding feature gate (`SupplementalGroupsPolicy`) is now enabled by default. This feature enables to implement more precise control over supplemental groups in containers that can strengthen the security posture, particularly in accessing volumes. Moreover, it also enhances the transparency of UID/GID details in Pod manifests, offering improved security oversight.
```

    supplementalGroups: [4000]
  containers:
  - name: ctr
    image: registry.k8s.io/e2e-test-images/agnhost:2.45
    command: [ "sh", "-c", "sleep 1h" ]
    securityContext:
      allowPrivilegeEscalation: false
```

What is the result of `id` command in the `ctr` container? The output should be similar to this:

```none
uid=1000 gid=3000 groups=3000,4000,50000
```

Where does group ID `50000` in supplementary groups (`groups` field) come from, even though `50000` is not defined in the Pod's manifest at all? The answer is `/etc/group` file in the container image.

Checking the contents of `/etc/group` in the container image should show below:

```none
user-defined-in-image:x:1000:
group-defined-in-image:x:50000:user-defined-in-image
```

Aha! The container's primary user `1000` belongs to the group `50000` in the last entry.

Thus, the group membership defined in `/etc/group` in the container image for the container's primary user is _implicitly_ merged to the information from the Pod. Please note that this was a design decision the current CRI implementations inherited from Docker, and the community never really reconsidered it until now.

### What's wrong with it?

> The _implicitly_ merged group information from `/etc/group` in the container image may cause some concerns particularly in accessing volumes (see [kubernetes/kubernetes#112879](https://issue.k8s.io/112879) for details) because file permission is controlled by uid/gid in Linux. Even worse, the implicit gids from `/etc/group` can not be detected/validated by any policy engines because there is no clue for the implicit group information in the manifest. This can also be a concern for Kubernetes security.

Reviewer: I think this could do a bit of sentence structure update, to highlight the security risk earlier, and then explain more in detail later. Also, UID and GID are usually in capital case. (This follows the same pattern used in https://github.com/kubernetes/website/pull/49983 for user namespace blog post.)

```suggestion
The _implicitly_ merged group information from `/etc/group` in the container image poses a security risk. These implicit GIDs can't be detected or validated by policy engines because there's no record of them in the Pod manifest. This can lead to unexpected access control issues, particularly when accessing volumes (see [kubernetes/kubernetes#112879](https://issue.k8s.io/112879) for details) because file permission is controlled by UID/GIDs in Linux.
```

## Fine-grained supplemental groups control in a Pod: `supplementaryGroupsPolicy`

> To tackle the above problem, Pod's `.spec.securityContext` now have `supplementalGroupsPolicy` field.

Reviewer: ```suggestion
To tackle the above problem, Pod's `.spec.securityContext` now includes `supplementalGroupsPolicy` field.
```

> This field provides a way to control how to calculate supplementary groups for the container processes in a Pod. The available policy is below:

Reviewer: Slight wording adjustments to make it a bit easier to follow

```suggestion
This field lets you control how Kubernetes calculates the supplementary groups for container processes within a Pod. The available policies are:
```

* _Strict_: it only attaches specified group IDs in `fsGroup`, `supplementalGroups`, or `runAsGroup` fields as the supplementary groups of the container processes. This means no group membership defined in `/etc/group` for the container's primary user will be merged.

Let's see how `Strict` policy works. Below Pod manifest specifies `supplementalGroupsPolicy: Strict`:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: strict-supplementalgroups-policy
spec:
  securityContext:
    runAsUser: 1000
    runAsGroup: 3000
    supplementalGroups: [4000]
    supplementalGroupsPolicy: Strict
  containers:
  - name: ctr
    image: registry.k8s.io/e2e-test-images/agnhost:2.45
    command: [ "sh", "-c", "sleep 1h" ]
    securityContext:
      allowPrivilegeEscalation: false
```

The result of `id` command in the `ctr` container should be similar to this:

```none
uid=1000 gid=3000 groups=3000,4000
```

You can see `Strict` policy can exclude group `50000` from `groups`! 

Thus, ensuring `supplementalGroupsPolicy: Strict` (enforced by some policy mechanism) helps prevent the implicit supplementary groups in a Pod.

{{<note>}}

> Actually, this is not enough because container with sufficient privileges / capability can change its process identity. Please see the following section for details.

Reviewer: Update the wording to start with the main point, provide some context, and then explain what comes next -- which I find is easier to follow personally. What do you think?

```suggestion
Keep in mind: a container with sufficient privileges can still change its process identity. The `supplementalGroupsPolicy` only affect the initial process identity. See the following section for details.
```

{{</note>}}

## `Strict` Policy requires newer CRI versions

Actually, CRI runtime (e.g. containerd, CRI-O) plays a core role for calculating supplementary group ids to be attached to the containers. Thus, `SupplementalGroupsPolicy=Strict` requires a CRI runtime that support this feature (`SupplementalGroupsPolicy: Merge` can work with the CRI runtime which does not support this feature because this policy is fully backward compatible policy).

Here are some CRI runtimes that support this feature, and the versions you need
to be running:

- containerd: v2.0 or later
- CRI-O: v1.31 or later

And, you can see if the feature is supported in the Node's `.status.features.supplementalGroupsPolicy` field.

```yaml
apiVersion: v1
kind: Node
...
status:
  features:
    supplementalGroupsPolicy: true
```

## The Behavioral Changes Introduced In Beta

> At alpha release, when a pod with `SupplementalGroupsPolicy=Strict` are scheduled to a node that does NOT support this feature(i.e. `.status.features.supplementalGroupsPolicy=false`), the pod's supplemental groups policy falls back to the `Merge` policy _silently_.

Reviewer: Some minor grammar fix

```suggestion
In the alpha release, when a Pod with `supplementalGroupsPolicy: Strict` was scheduled to a node that did not support the feature (i.e., `.status.features.supplementalGroupsPolicy=false`), the Pod's supplemental groups policy silently fell back to `Merge`.
```

> However, since the beta release (v1.33), to enforce the policy more strictly, __such pod creation will be rejected by kubelet because the node can not ensure the specified policy__. When your pod is rejected, you will see warning events with `reason=SupplementalGroupsPolicyNotSupported` like below:

Reviewer: Because this is a blog about v1.33 release, it could emphasise that a bit more? (I took out the markdown emphasis to simplify for now, please add as you see fit!)

```suggestion
In v1.33, this has entered beta to enforce the policy more strictly, where kubelet rejects pods whose nodes cannot ensure the specified policy. If your pod is rejected, you will see warning events with `reason=SupplementalGroupsPolicyNotSupported` like below:
```

```yaml
apiVersion: v1
kind: Event
...
type: Warning
reason: SupplementalGroupsPolicyNotSupported
message: "SupplementalGroupsPolicy=Strict is not supported in this node"
involvedObject:
  apiVersion: v1
  kind: Pod
  ...
```

## Upgrade Consideration

> If you already used the feature, particularly `SupplementalGroupsPolicy=Strict` policy, we assume your cluster's CRI runtimes already support this feature. In this case, you do not need to worry about the pod rejections described in the above section.
> 
> However, if your cluster:
> 
> - use `SupplementalGroupsPolicy=Strict` policy, but
> - its CRI runtimes do NOT support this feature yet(`.status.features.supplementalGroupsPolicy=false`),

Reviewer: Just minor wording update

```suggestion
If you're already using this feature, especially the `supplementalGroupsPolicy: Strict` policy, we assume that your cluster's CRI runtimes already support this feature. In that case, you don't need to worry about the pod rejections described above.

However, if your cluster:

- uses the `supplementalGroupsPolicy: Strict` policy, but
- its CRI runtimes do NOT yet support the feature (i.e., `.status.features.supplementalGroupsPolicy=false`),
```
